File: 502-ipo/ipo.py
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        profits_heap = [] # max heap
        capital_heap = [] #min heap

        for i in range(0,len(capital)):
            heapq.heappush(capital_heap, (capital[i], i))
        
        
        for _ in range(k):
            while capital_heap and capital_heap[0][0] <= w:
                c,i = heapq.heappop(capital_heap)
                heapq.heappush(profits_heap, (-profits[i], i))

            if not profits_heap:
                break
            
            profit, _ = heappop(profits_heap)
            w += -profit
        return w

File: 502-ipo/ipo.java
class Solution {
    class Pair{
        int profit;
        int capital;
        Pair(int profit, int capital){
            this.profit = profit;
            this.capital = capital;
        }
    }
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        PriorityQueue<Pair> minHeap = new PriorityQueue<>(new Comparator<Pair>(){
            public int compare(Pair p1, Pair p2){
                return Integer.compare(p1.capital, p2.capital);
            }
        });
        PriorityQueue<Pair> maxHeap = new PriorityQueue<>(new Comparator<Pair>(){
            public int compare(Pair p1, Pair p2){
                return Integer.compare(p2.profit, p1.profit);
            }
        });

        for(int i=0;i<profits.length;i++){
            minHeap.add(new Pair(profits[i], capital[i]));
        }

        while(k>0){
            while(!minHeap.isEmpty() && minHeap.peek().capital<=w){
                maxHeap.add(minHeap.remove());
            }
            if(maxHeap.isEmpty()) break;
            w += maxHeap.remove().profit;
            k--;
        }
        return w;
    }
}